src/ray/raylet/scheduling/cluster_task_manager.cc

```diff
@@ -246,38 +239,81 @@ void ClusterTaskManager::DispatchScheduledTasksToWorkers(
               cluster_resource_scheduler_->SerializedTaskResourceInstances(
                   allocated_instances);
         }
-        std::shared_ptr<WorkerInterface> worker =
-            worker_pool_.PopWorker(spec, allocated_instances_serialized_json);
-        if (!worker) {
-          RAY_LOG(DEBUG) << "This node has available resources, but no worker processes "
-                            "to grant the lease.";
-          // We've already acquired resources so we need to release them to avoid
-          // double-acquiring when the next invocation of this function tries to schedule
-          // this task.
-          cluster_resource_scheduler_->ReleaseWorkerResources(allocated_instances);
-          ReleaseTaskArgs(task_id);
-          // It may be that no worker was available with the correct runtime env or
-          // correct job ID.  However, another task with a different env or job ID
-          // might have a worker available, so continue iterating through the queue.
-          work_it++;
-          // Keep track of runtime env that doesn't have workers available so we
-          // won't call PopWorker for subsequent tasks requiring the same runtime env.
-          blocked_runtime_env_to_skip.insert(runtime_env_worker_key);
-          continue;
-        }
-
-        RAY_LOG(DEBUG) << "Dispatching task " << task_id << " to worker "
-                       << worker->WorkerId();
-        auto reply = std::get<1>(*work_it);
-        auto callback = std::get<2>(*work_it);
-        Dispatch(worker, leased_workers_, allocated_instances, task, reply, callback);
-      }
-
-      if (!spec.GetDependencies().empty()) {
-        task_dependency_manager_.RemoveTaskDependencies(
-            task.GetTaskSpecification().TaskId());
+        work->status = WorkStatus::WAITING_FOR_WORKER;
+        worker_pool_.PopWorker(
+            spec,
+            [this, allocated_instances, task_id, scheduling_class, work](
+                const std::shared_ptr<WorkerInterface> worker, Status status) -> bool {
+              const auto &reply = work->reply;
```

Comment:
Suggest making this lambda a method of the ClusterTaskManager class (just easier to read).

Reply:
I have a lot variables captured. And lambda is used a lot in our project. Maybe just leave it a lambda ok?